funkwhale: init at 0.20.0

mmai opened the pull request:
###### Motivation for this change

This adds a new package and module which enable the Funkwhale music web service  (https://funkwhale.audio/).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Infinisil commented on nixos/modules/services/web-apps/funkwhale/requirements.nix:
```diff
@@ -0,0 +1,1788 @@
+
+    "rsa" = python.mkDerivation {
+      name = "rsa-4.0";
+      src = pkgs.fetchurl { url = "https://files.pythonhosted.org/packages/cb/d0/8f99b91432a60ca4b1cd478fd0bdf28c1901c58e3a9f14f4ba3dba86b57f/rsa-4.0.tar.gz"; sha256 = "1a83640640573
```
Read https://nixos.org/nixpkgs/manual/#contributing on how to contribute python packages. pypi2nix isn't desired because of these large files. Take a look at other python packages.

Infinisil commented on pkgs/servers/web-apps/funkwhale/default.nix:
```diff
@@ -0,0 +1,42 @@
+{ pkgs ? import <nixpkgs> {} }:
+with pkgs;
```
`import <nixpkgs> {}` is disallowed in nixpkgs because it's impure. Take dependencies as arguments, look at the other packages and read up on callPackage.

Infinisil commented on pkgs/servers/web-apps/funkwhale/0001-changes.patch:
```diff
@@ -0,0 +1,31 @@
++env.read_env(env_file=env.str('FUNKWHALE_ENV_FILE', '.env'))
+ 
+ FUNKWHALE_HOSTNAME = None
+ FUNKWHALE_HOSTNAME_SUFFIX = env("FUNKWHALE_HOSTNAME_SUFFIX", default=None)
```
Can you make this patch retain the original behavior if the nix-specific `FUNKWHALE_ENV_FILE` is empty?

mmai replied:
No need for a patch anymore with the funkwhale 0.18 version

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+  };
+
+  meta = {
+    maintainers = with lib.maintainers; [ mmai ];
```
Fill in the rest of the meta info, look at other packages.

mmai replied:
You mean the "doc" option ? There is nothing else I can put for a module according to https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/misc/meta.nix. The meta info in pkgs/servers/web-apps/funkwhale/default.nix are correctly set I think.

Infinisil replied:
Oh sorry, I mixed it up with the package declaration, my fault.

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+          environment = funkwhaleEnv;
+          script = ''
+            if ! test -e ${funkwhaleHome}/run; then
+            mkdir -p ${funkwhaleHome}/run
```
Check out systemd's `RuntimeDirectory` option, no need to do this manually.

mmai replied:
Thanks. It's fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.env.nix:
```diff
@@ -0,0 +1,149 @@
+
+# Nginx related configuration
+NGINX_MAX_BODY_SIZE=30M
+''
```
The explanatory comments aren't very useful if the user can't modify anything anyways through it. The only interface you provide is the NixOS service options, and users can set those instead. So you can just remove all the comments (and irrelevant sections, like the docker ones). Then you can also just inline this file.

mmai replied:
thanks. There's only one file now, and such comments are removed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+    services.funkwhale = {
+      enable = mkEnableOption "funkwhale";
+
+      api_ip = mkOption {
```
NixOS options use camelCase.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+      };
+
+      api_port = mkOption {
+        type = types.str;
```
Use `types.port`

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+        type = types.str;
+        default = "127.0.0.1";
+        description = ''
+          Funkwhale API IP
```
All description fields should end with a `.`

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+      };
+
+      protocol = mkOption {
+        type = types.str;
```
`types.enum [ "http" "https" ]`

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+        default = "consolemail://";
+        description = ''
+          Configure email sending using this variale
+          By default, funkwhale will output emails sent to stdout
```
What does `will output emails sent to stdout` mean?

mmai replied:
it outputs emails to the console instead of sending them. I've added a link to the online doc.

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+          Configure email sending using this variale
+          By default, funkwhale will output emails sent to stdout
+          here are a few examples for this setting
+          consolemail://         # output emails to console (the default)
```
Enclose this in either a `<literal>` (keyword) or `<programlisting>` (multiline) xml section. Build the docs yourself to find out whether they render correctly.

mmai replied:
thanks. fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+          On a production instance, you'll usually want to use an external SMTP server:
+          smtp://user@:password@youremail.host:25
+          smtp+ssl://user@:password@youremail.host:465
+          smtp+tls://user@:password@youremail.host:587
```
Can you link to the upstream docs here?

mmai replied:
yes. I added the link

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+
+      default_from_email = mkOption {
+        type = types.str;
+        default = "noreply@yourdomain";
```
This shouldn't be a default.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+
+      hostname = mkOption {
+        type = types.str;
+        default = "yourdomain";
```
This shouldn't be a default.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+      #     the API will use different kind of headers to serve audio files
+      #     Allowed values: nginx, apache2
+      #   '';
+      # };
```
Needs to be removed.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+
+      django_allowed_hosts = mkOption {
+        type = types.str;
+        default = "yourdomain";
```
Shouldn't be a default.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+
+      django_secret_key = mkOption {
+        type = types.str;
+        default = "";
```
Shouldn't be a default.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+      recommendedTlsSettings = true;
+      recommendedOptimisation = true;
+      recommendedGzipSettings = true;
+      recommendedProxySettings = true;
```
A service shouldn't set these options on behalf of the user.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+        "${cfg.hostname}" = {
+        # enableACME = true; #Ask Let's Encrypt to sign a certificate for this vhost
+        # addSSL = true;
+        # forceSSL = true;
```
Remove all other such comments.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+              extraConfig = ''
+                internal;
+              '';
+              alias = "${cfg.api.media_root}";
```
All such options can be set directly: `alias = cfg.api.media_root;`.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale/funkwhale.nix:
```diff
@@ -0,0 +1,329 @@
+            if ! test -e ${cfg.api.media_root}; then
+            mkdir -p ${cfg.api.media_root}
+            mkdir -p ${cfg.api.static_root}
+            mkdir -p ${cfg.music_directory_path}
```
Look into `systemd.tmpfiles.rules` to create these directories for you.

mmai replied:
thanks. fixed.

infinisil reviewed (changes requested):
All in all, this PR is far from done and needs a lot of work. Note that a lot of the comments I left also apply to many other places in your changes.

matthiasbeyer commented:
Any progress here?

ivan commented:
For the author, reviewers, and committers: this PR was scanned and appears to add a use of the deprecated `types.string`, which emits a warning as of #66346. Before merging, please change this to another type, possibly:

- `types.str` for a single string where merging does not make sense, or cannot work
- `types.lines` for multi-line configuration or scripts where merging is possible
- `types.listOf types.str` for a mergeable list of strings

FRidh commented:
Nothing happening so closing.

mmai commented:
@FRidh I'm not sure I understand. Do you expect more actions from me ? I've made all the requested changes for the initial pull request but my fixes didn't received any review.

FRidh commented on pkgs/development/python-modules/django-auth-ldap/default.nix:
```diff
@@ -0,0 +1,31 @@
+    sha256 = "72848b3b036d299114be3c6ef38b12f83f6cf1cf1696c5f92e06fe45a1b8e27b";
+  };
+
+  buildInputs = [ ldap django_2_2 ]; 
```
python packages need to be added as `propagatedBuildInputs`

mmai replied:
fixed

FRidh commented on pkgs/development/python-modules/django-auth-ldap/default.nix:
```diff
@@ -0,0 +1,31 @@
+    license = licenses.bsd2;
+    maintainers = with maintainers; [ mmai ];
+    broken = stdenv.isDarwin;
+    platforms = platforms.unix;
```
this is already being set by `buildPython*`

mmai replied:
removed the `plateforms = platforms.unix` line

FRidh commented on pkgs/servers/web-apps/funkwhale/default.nix:
```diff
@@ -0,0 +1,40 @@
+  version = "${release}";
+  src = srcs.api;
+  buildInputs = [ unzip ];
+  propagatedBuildInputs = [ ];
```
empty so remove

mmai replied:
fixed

FRidh commented on pkgs/servers/web-apps/funkwhale/default.nix:
```diff
@@ -0,0 +1,40 @@
+  name = "funkwhale";
+  version = "${release}";
+  src = srcs.api;
+  buildInputs = [ unzip ];
```
nativeBuildInputs

mmai replied:
fixed

FRidh commented:
Please do not merge master. Instead, rebase your changes on top of master.

mmai commented:
Thanks for the review. I also fixed the deprecated `types.string` and updated to funkwhale 0.19.1.

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+          wantedBy = [ "funkwhale-server.service" "funkwhale-worker.service" "funkwhale-beat.service" ];
+          before   = [ "funkwhale-server.service" "funkwhale-worker.service" "funkwhale-beat.service" ];
+          environment = funkwhaleEnv;
+          script = ''
```
This is running as root as of now, this should be avoided.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+              set -a
+              . ${funkwhaleEnvFile}
+              set +a
+              ${pythonEnv}/bin/python ${pkgs.funkwhale}/manage.py createsuperuser" > ${cfg.dataDir}/createSuperUser.sh
```
What does this do and why? If you need to have environment variables set then set them before calling the executable, not like this.

mmai replied:
ok

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+
+          serviceConfig = serviceConfig;
+          environment = funkwhaleEnv;
+          script = "${pythonEnv}/bin/daphne -b ${cfg.apiIp} -p ${toString cfg.apiPort} config.asgi:application --proxy-headers";
```
Don't use `script` when it's just a single command, use `serviceConfig.ExecStart` instead, same for the other services.

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+
+      };
+
+      environment.systemPackages = with pkgs; [ ffmpeg libjpeg ];
```
Why this?

mmai replied:
These dependencies don't seem to be required anymore, so i've removed this line.
(ffmpeg was used to stream music, libjpeg to manage albums covers. Should  I have required them in the funkwhale package instead of the module ?)

Infinisil commented on maintainers/maintainer-list.nix:
```diff
@@ -4199,6 +4199,11 @@
   };
+  mmai = {
+    email = "henri.bourcereau@gmail.com";
+    github = "mmai";
```
Since https://github.com/NixOS/nixpkgs/pull/66361 you need to provide the githubId, see comment at the top of this file

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+    DEFAULT_FROM_EMAIL=${cfg.defaultFromEmail}
+    REVERSE_PROXY_TYPE=nginx
+    DATABASE_URL=postgresql://funkwhale@:5432/funkwhale
+    CACHE_URL=redis://127.0.0.1:6379/0
```
redis' port comes from `config.services.redis.port`

mmai replied:
fixed

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+      enable = true;
+      authentication = pkgs.lib.mkOverride 10 ''
+        local all all trust
+        host all all ::1/128 trust
```
And this here also disables authentication for local connections. This can't stay either.

mmai replied:
fixed

infinisil reviewed (changes requested):
There's still so many things to correct. The NixOS module especially is in a state impossible to accept, and I probably missed some stuff too.

worldofpeace commented on pkgs/development/python-modules/unicode-slugify/default.nix:
```diff
@@ -0,0 +1,29 @@
+  };
+
+  meta = with stdenv.lib; {
+    description = "Unicode Slugify is a slugifier that generates unicode slugs. It was originally used in the Firefox Add-ons web site to generate slugs for add-ons and add-on collections. Many of these add-ons and collections had unicode characters and required more than simple transliteration.";
```
This description is too long. Can we shorten it and not use a period at the end?

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/unicode-slugify/default.nix:
```diff
@@ -0,0 +1,29 @@
+  pname = "unicode-slugify";
+  version = "0.1.3";
+
+  propagatedBuildInputs = [ unidecode ];
```
six should be here https://github.com/mozilla/unicode-slugify/blob/master/setup.py#L22

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/pymemoize/default.nix:
```diff
@@ -0,0 +1,25 @@
+}:
+
+buildPythonPackage rec {
+  pname = "PyMemoize";
```
Please normalize `pname` like `pymemoize`.
Note that the url for pypi includes `PyMemoize` so you'll have to do
```
pname = "PyMemoize";
```
in the `fetchPypi` call.

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/pymemoize/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+  meta = with stdenv.lib; {
+    description = "Simple Python cache and memoizing module";
+    homepage = "http://github.com/mikeboers/PyMemoize";
```
```suggestion
    homepage = "https://github.com/mikeboers/PyMemoize";
```

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/hiredis/default.nix:
```diff
@@ -0,0 +1,27 @@
+  doCheck = false;
+
+  meta = with stdenv.lib; {
+    description = "Python extension that wraps protocol parsing code in hiredis. It primarily speeds up parsing of multi bulk replies";
```
This description is too long. Can we shorten it and not use a period at the end?

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/django-versatileimagefield/default.nix:
```diff
@@ -0,0 +1,27 @@
+  doCheck = false;
+
+  meta = with stdenv.lib; {
+    description = "A drop-in replacement for django's ImageField that provides a flexible, intuitive and easily-extensible interface for creating new images from the one assigned to the field";
```
This description is too long. Can we shorten it and not use a period at the end?

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/django-versatileimagefield/default.nix:
```diff
@@ -0,0 +1,27 @@
+
+  meta = with stdenv.lib; {
+    description = "A drop-in replacement for django's ImageField that provides a flexible, intuitive and easily-extensible interface for creating new images from the one assigned to the field";
+    homepage = "http://github.com/respondcreate/django-versatileimagefield/";
```
```suggestion
    homepage = "https://github.com/respondcreate/django-versatileimagefield/";
```

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/django-filter/default.nix:
```diff
@@ -0,0 +1,34 @@
+  '';
+
+  meta = with stdenv.lib; {
+    description = "Django-filter is a reusable Django application for allowing users to filter querysets dynamically";
```
```suggestion
    description = "Reusable Django application for allowing users to filter querysets dynamically";
```

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/django-filter/default.nix:
```diff
@@ -0,0 +1,34 @@
+
+buildPythonPackage rec {
+  pname = "django-filter";
+  version = "2.1.0";
```
Out of date
https://github.com/carltongibson/django-filter/releases/tag/2.2.0

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/django-auth-ldap/default.nix:
```diff
@@ -0,0 +1,30 @@
+buildPythonPackage rec {
+  pname = "django-auth-ldap";
+  version = "1.7.0";
+
```
```suggestion
  disabled = isPy27;
```

This doesn't support that version.

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/django-auth-ldap/default.nix:
```diff
@@ -0,0 +1,30 @@
+
+buildPythonPackage rec {
+  pname = "django-auth-ldap";
+  version = "1.7.0";
```
out of date https://github.com/django-auth-ldap/django-auth-ldap/releases/tag/2.0.0

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/django-auth-ldap/default.nix:
```diff
@@ -0,0 +1,30 @@
+  doCheck = false;
+
+  meta = with stdenv.lib; {
+    description = "Django LDAP authentication backend";
```
```suggestion
    description = "Django authentication backend that authenticates against an LDAP service";
```

mmai replied:
fixed

worldofpeace commented on pkgs/development/python-modules/django-auth-ldap/default.nix:
```diff
@@ -0,0 +1,30 @@
+    homepage = https://github.com/django-auth-ldap/django-auth-ldap;
+    license = licenses.bsd2;
+    maintainers = with maintainers; [ mmai ];
+    broken = stdenv.isDarwin;
```
does this even support darwin? If it doesn't it would make more sense to set platforms
```suggestion
    platforms = platforms.linux
```

mmai replied:
fixed

worldofpeace commented on pkgs/top-level/python-packages.nix:
```diff
@@ -2920,6 +2924,8 @@ in {
 
   django-sampledatahelper = callPackage ../development/python-modules/django-sampledatahelper { };
 
+	django-versatileimagefield = callPackage ../development/python-modules/django-versatileimagefield  { };
```
Indentation is misaligned here.

mmai replied:
fixed

worldofpeace reviewed (changes requested):
Hi :wave: 

I've reviewed the python modules. I did notice that a few of them were out of date.
If this becomes a long running PR I'm available to merge the modules separately, as the python framework in nixpkgs moves quickly. So we tend to get duplicated expressions over time.

Additionally I noticed that the hashes for those packages are in base64, and we'd like python packages to have base32 hashes. If possible if would be preferred if you could convert the hashes with
```
nix to-base32 $hash

# or
nix-hash --to-base32 --type sha256 $hash
```

Thanks ✨️

mmai committed:
funkwhale: init at 0.20.0

mmai committed:
nixos/funkwhale: init

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+
+			# Needs
+			#  pillow 5.4 ?
+			#  mutagen 1.42 ?
```
fixed

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+let
+  pythonEnv = (pkgs.python3.override {
+    packageOverrides = self: super: rec {
+      django = super.django_2_2;
```
fixed

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+			#  mutagen 1.42 ?
+
+      # celery package tests fail
+      celery = super.celery.overridePythonAttrs(old: rec { doCheck = false; });
```
fixed for the two packages

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+
+      # needs PyLD > 1.0
+      PyLD = super.buildPythonPackage rec {
+        pname = "PyLD";
```
done

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+
+      # 2.1.0 needed (last version on pypi = 3.0.1)
+      django-cleanup = super.buildPythonPackage rec {
+        pname = "django-cleanup";
```
fixed

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+        };
+        checkInputs = [ super.nose ];
+        checkPhase = "nosetests";
+      };
```
done

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+      persisting-theory = super.buildPythonPackage rec {
+        pname = "persisting-theory";
+        version = "0.2.1";
+        name = "${pname}-${version}";
```
fixed

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+
+    users.groups.funkwhale = {};
+
+    services.ntp.enable = true; 
```
line removed

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+    services.postgresql = {
+      enable = true;
+      authentication = pkgs.lib.mkOverride 10 ''
+        local all all trust
```
Sorry about that. fixed

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+      initialScript = pkgs.writeText "backend-initScript" ''
+        CREATE ROLE funkwhale WITH LOGIN PASSWORD 'funkwhale' CREATEDB;
+        CREATE DATABASE funkwhale WITH ENCODING 'utf8';
+        GRANT ALL PRIVILEGES ON DATABASE funkwhale TO funkwhale;
```
fixed

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+            EnvironmentFile =  "${funkwhaleEnvFile}";
+          };
+        in {
+        "funkwhale.target" = {
```
fixed

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+    EMAIL_CONFIG=${cfg.emailConfig}
+    DEFAULT_FROM_EMAIL=${cfg.defaultFromEmail}
+    REVERSE_PROXY_TYPE=nginx
+    DATABASE_URL=postgresql://funkwhale@:5432/funkwhale
```
fixed

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+            User = "postgres";
+          };
+          script = ''
+            ${pkgs.postgresql}/bin/psql -d funkwhale -c 'CREATE EXTENSION IF NOT EXISTS "unaccent";'
```
fixed

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+
+      django-oauth-toolkit = super.buildPythonPackage rec {
+        pname = "django-oauth-toolkit";
+        version = "1.2.0";
```
fixed

mmai commented:
Hi @worldofpeace I've fixed the descriptions and the hashes of the python packages. There are also new packages extracted from the funkwhale NixOs module. You will see that I've disabled the check phase for half of them, I didn't find obvious ways to make tests pass for those, and my python abilities are limited. If you want to merge the python packages separately, should I prepare distinct pull requests ?

mmai commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,565 @@
+    DJANGO_SETTINGS_MODULE=config.settings.production
+    DJANGO_SECRET_KEY=${cfg.api.djangoSecretKey}
+    RAVEN_ENABLED=${boolToString cfg.enableRaven}
+    RAVEN_DSN=https://44332e9fdd3d42879c7d35bf8562c6a4:0062dc16a22b41679cd5765e5342f716@sentry.eliotberriot.com/5
```
This is the funkwhale developpers Raven DSN used to track anomalies. I've converted this to a config parameter with this value as the default.

worldofpeace commented:
@mmai I've pushed some fixup commits here for things I knew how to fix right away.
I think I can do one last quick review of the modules, and I can just push them to master from this PR.

worldofpeace commented on pkgs/development/python-modules/channels-redis/default.nix:
```diff
@@ -0,0 +1,26 @@
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = http://github.com/django/channels_redis/;
```
https

worldofpeace commented on pkgs/development/python-modules/channels-redis/default.nix:
```diff
@@ -0,0 +1,26 @@
+    pname = "channels_redis";
+    sha256 = "1g4izdf8237pwxn85bv5igc2bajrvck1p2a7q448qmjfznrbrk5p";
+  };
+  buildInputs = [ redis channels msgpack aioredis hiredis asgiref ];
```
The following install requires should be `propagatedBuildInputs` 
* https://github.com/django/channels_redis/blob/master/setup.py#L34

worldofpeace commented on pkgs/development/python-modules/django-dynamic-preferences/default.nix:
```diff
@@ -0,0 +1,21 @@
+    inherit pname version;
+    sha256 = "1z2dndkpypk4pvb0byh5a771vgp50vn8g1rbk5r3sml6dm4wcn7s";
+  };
+  buildInputs = [ six django persisting-theory ]; 
```
These should all be `propagatedBuildInputs`
* https://github.com/EliotBerriot/django-dynamic-preferences/blob/develop/setup.py#L39

worldofpeace commented on pkgs/development/python-modules/django-oauth-toolkit/default.nix:
```diff
@@ -0,0 +1,26 @@
+buildPythonPackage rec {
+  pname = "django-oauth-toolkit";
+  version = "1.2.0";
+  name = "${pname}-${version}";
```
```suggestion
```

worldofpeace commented on pkgs/development/python-modules/django-oauth-toolkit/default.nix:
```diff
@@ -0,0 +1,26 @@
+    rev = version;
+    sha256 = "1zbksxrcxlqnapmlvx4rgvpqc4plgnq0xnf45cjwzwi1626zs8g6";
+  };
+  buildInputs = [ django_2_2 requests oauthlib ];
```
These should be `propagatedBuildInputs`.

worldofpeace commented on pkgs/development/python-modules/django-storages/default.nix:
```diff
@@ -0,0 +1,25 @@
+    sha256 = "01xq232h321716r08rari9payas7fsiwwr5q6zgcrlwkckwxxczk";
+  };
+
+  buildInputs = [ django ];
```
This should be `propagateBuildInputs`
* https://github.com/jschneier/django-storages/blob/master/setup.py#L16

worldofpeace commented on pkgs/development/python-modules/persisting-theory/default.nix:
```diff
@@ -0,0 +1,24 @@
+  checkPhase = "nosetests";
+
+  meta = with stdenv.lib; {
+    homepage = http://code.eliotberriot.com/eliotberriot/persisting-theory;
```
https

worldofpeace reviewed (changes requested):
* `pythonPackages.channels-redis`

Please add
```
disabled = pythonOlder "3.6";
```
for https://github.com/django/channels_redis/blob/c436cca4fccd0d2cf02b4faac7d0927710b3c7f1/setup.py#L32

Possibly also setup tests
* https://github.com/django/channels_redis/blob/master/setup.py#L13
which use `pytest`.

Additionally, I'm thinking there might be some confusion on what `django` is being used in the python modules. `django` in the python package set will always be the python module, where perhaps there's a requirement for `pkgs.django`?

Other than that, see my fixup commits for correct whitespace when I've corrected it.

mmai commented:
@worldofpeace I've rebased the fixups. 

I've removed the explicit dependencies on django_2_2 on django-oauth-toolkit and django-auth-ldap. django-auth-ldap is in fact compatible with django 1, and for django-oauth-toolkit I took inspiration on django-raster to check the django version.

I tried to setup tests on channels-redis with the github sources, but they eventually fail, I've kept the code in comments.

worldofpeace commented:
@mmai I'll see about integrating the python packages later today.

worldofpeace commented:
Lol more like today now :rofl:

worldofpeace commented:
Merged all python modules in #71262. 

We need to drop all those commits from here, and the funkwhale commit needs a rewrite to not add you to the maintainer list. I had to do that separately for merging the modules https://github.com/NixOS/nixpkgs/pull/71262/commits/3895e097c365fee76149efd676d94bef5952bdf2.

mmai commented:
Thanks @worldofpeace . I've removed the python packages commits.

Infinisil commented on pkgs/servers/web-apps/funkwhale/default.nix:
```diff
@@ -0,0 +1,46 @@
+  installPhase = ''
+    mkdir $out
+    cp -R ./* $out
+    unzip ${srcs.frontend} -d $out
```
In nixpkgs there's a structure for $out you need to follow. And since this is a python package, you should use `buildPythonPackage` to build, which will do a lot of things automatically. See section 10.15.2.2.2 in the [Nixpkgs manual](https://nixos.org/nixpkgs/manual/#building-packages-and-applications). This way you can also specify the dependencies in the package directly.

FRidh replied:
This is a Django application and they usually do not provide any installer. While it would be nice if funkwhale would be usable without the NixOS module, given its a Django app it is typically managed using the `manage.py` script and various other executables are executed (gunicorn e.g.). Therefore, I think this approach is alright.

Note in another PR (https://github.com/NixOS/nixpkgs/pull/67951/files#diff-5e9f694ce07aa22a80eb0c049c03beaf) I recommended putting the contents of the Django application in a `site-packages` folder using `buildPythonPackage` because then the application could be added to the env (that would correspond to adding the `funkwhale` derivation to the Python env. That would be even nicer, especially if the Python env would be moved here!

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+          group = cfg.group;
+        });
+
+      users.groups = optionalAttrs (cfg.group == "funkwhale") (singleton { name = "funkwhale"; });
```
Do `users.groups.funkwhale = mkIf (cfg.group == "funkwhale") {}` instead, and similarly for `users.users`

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+          djangoSecretKey = mkOption {
+            type = types.str;
+            description = ''
+              Django secret key. Generate one using `openssl rand -base64 45` for example.
```
Docs aren't markdown but asciidoc, so here you need `<command>the command</command>` instead

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+            type = types.str;
+            default = "/srv/funkwhale/static";
+            description = ''
+              Where static files (such as API css or icons) should be compiled on your system ? Ensure this directory actually exists.
```
```suggestion
              Where static files (such as API css or icons) should be compiled on your system. Ensure this directory actually exists.
```

Same for other descriptions where there's a random question mark.

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+          type = types.str;
+          default = "consolemail://";
+          description = ''
+            Configure email sending. By default, it outputs emails to console instead of sending them. See https://docs.funkwhale.audio/configuration.html#email-config for details.
```
Syntax for links is `<link xlink:href="https://example.com"/>`

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+            type = types.str;
+            default = "/srv/funkwhale/media";
+            description = ''
+              Where media files (such as album covers or audio tracks) should be stored on your system ? Ensure this directory actually exists.
```
No need to write "Ensure it exists" since you use `tmpfiles.rules` to guarantee that already

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+            type = types.nullOr types.path;
+            default = "/run/postgresql";
+            defaultText = "/run/postgresql";
+            example = "/run/postgresql";
```
No need to specify `defaultText` when the `default` is the same. And no need to specify an example if it's the same as the default

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+          port = mkOption {
+            type = types.int;
+            default = 5432;
+            defaultText = "5432";
```
Also here, no need to set `defaultText`

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+      let serviceConfig = {
+        User = "${cfg.user}";
+        WorkingDirectory = "${pkgs.funkwhale}";
+        EnvironmentFile =  "${funkwhaleEnvFile}";
```
This shouldn't be necessary since you set `environment` in the services already

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+          };
+        };
+
+        musicDirectoryPath = mkOption {
```
Maybe call it `musicPath` instead

matthiasbeyer replied:
It is more consistent with the funkwhaleEnvironment variable name (which is `MUSIC_DIRECTORY_PATH`), so IMO it should not be changed, because of consistency.

Infinisil commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+          before   = [ "funkwhale-init.service" ];
+          serviceConfig = {
+            User = "postgres";
+            ExecStart = '' ${config.services.postgresql.package}/bin/psql -d ${cfg.database.name}  -c 'CREATE EXTENSION IF NOT EXISTS "unaccent";CREATE EXTENSION IF NOT EXISTS "citext";' '';
```
Split this into more lines, same for other long lines, try to keep them to ~80 chars wide

infinisil reviewed (changes requested):
Still some important things (and some less important things) to correct, but it's looking *much* better than initially!

FRidh commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,543 @@
+            Group = "${cfg.group}";
+          };
+          script = ''
+            ${pythonEnv}/bin/python ${pkgs.funkwhale}/manage.py migrate
```
`${python.interpreter}`

FRidh reviewed (commented):
Nice work. It's a very big application. I am worried though how we can keep it stable. I imagine that if the pythonEnv is moved to `pkgs.funkwhale`, that we can provide overrides there as needed. It might also be good to provide a NixOS test or lighter tests for the `pkgs.funkwhale` package so we know we don't break it. Maybe we should use the `requirements.txt` to verify deps as well?
Hook to test against a `requirements.txt` file. https://github.com/NixOS/nixpkgs/pull/71908

mmai commented:
@FRidh @Infinisil I didn't managed to set a structure resembling what you describe. Looking at the mailman package code and the documentation didn't help.

I'm really sorry to say that  at this point I am not able to work on this anymore. I feel that my python and nixos abilities are to shallow and I invested way more time than initially planned, and I see that we are far from having a stable package (the master branch already breaks python dependencies).

Feel free to close the pull request.

EliotBerriot commented:
Hi, I am one of Funkwhale's maintainer, and I wasn't aware of this packaging effort, so I'm a bit late. I don't have any knowledge about NixOS, however, I'd be willing to help if you need some clarification or tweaks on Funkwhale side to make it easier to packagers. 

I can also answer questions you have regarding our dependencies, for instance.

Many thanks to the people who worked on this!

davidak commented:
@matthiasbeyer would you be able to finish it?

matthiasbeyer commented:
Holy cow. I doubt it. Mainly because I do not know how to properly write services and also because I have little time. I will try next weekend though, if I do not forget.

matthiasbeyer commented:
I'm not sure how to push directly to this PR, nor am I convinced I should. I have some patches at https://github.com/matthiasbeyer/nixpkgs/tree/funkwhale though.

jbaum98 commented:
I'd also be happy to work on this. I've looked at the discussion, but can someone give a list of all the things that need to be changed or fixed to get this PR ready to merge?

worldofpeace commented:
@jbaum98  I believe https://github.com/NixOS/nixpkgs/pull/53416#pullrequestreview-306554871 and maybe https://github.com/NixOS/nixpkgs/pull/53416/files#r338540545?

You should re PR and review can begin again.

worldofpeace commented:
We also have https://github.com/NixOS/nixpkgs/pull/53416#issuecomment-573788340 who responsive to questions so this can be transplanted into NixOS as easily as possible.
But honestly, a lot of the issues here will be ecosystem ones.

worldofpeace commented:
I'm going to close this since @mmai said they won't be working on it anymore, so it won't be merged. The code should still be available though.

jbaum98 commented:
@FRidh I looked at the mailman PR, and I'm not quite clear on what structure we want for the Django app.

Also, this project has a pretty complicated architecture:
>The project relies on the following components and services to work:
> - A web application server (Python/Django/Gunicorn)
> - A PostgreSQL database to store application data
> - A redis server to store cache and tasks data
> - A celery worker to run asynchronous tasks (such as music import)
> - A celery scheduler to run recurrent tasks
> - A ntp-synced clock to ensure federation is working seamlessly
It seems like we want package the API server separately from the frontend server, and then hook that together with all the other stuff in the NixOS module.
